packages/shared-ux/chrome/navigation/src/model/create_side_nav.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { EuiSideNavItemType } from '@elastic/eui';
+import type { NavigationModelDeps } from '.';
+import type { NavItemProps, PlatformSectionConfig } from '../../types';
+import type { Locator } from '../../types/internal';
+
+type MyEuiSideNavItem = EuiSideNavItemType<unknown>;
+type OnClickFn = MyEuiSideNavItem['onClick'];
+
+/**
+ * Factory function to return a function that processes modeled nav items into EuiSideNavItemType
+ * The factory puts memoized function arguments in scope for iterations of the recursive item processing.
+ */
+export const createSideNavDataFactory = (
+  deps: NavigationModelDeps,
+  activeNavItemId: string | undefined
+) => {
+  const { basePath, navigateToUrl, getLocator, registerNavItemClick } = deps;
+  const createSideNavData = (
+    parentIds: string | number = '',
+    navItems: NavItemProps[],
+    platformSectionConfig?: PlatformSectionConfig
+  ): Array<EuiSideNavItemType<unknown>> =>
+    navItems.reduce<MyEuiSideNavItem[]>((accum, item) => {
+      const { id, name, items: subNav, locator: locatorDefinition, href } = item;
+      const config = platformSectionConfig?.properties?.[id];
+      if (config?.enabled === false) {
+        // return accumulated set without the item that is not enabled
+        return accum;
+      }
+
+      const { id: locatorId, params: locatorParams } = locatorDefinition ?? {};
+      let locator: Locator | undefined;
+      if (locatorId) {
+        locator = getLocator(locatorId);
+      }
+
+      let onClick: OnClickFn | undefined;
+
+      if (locatorId && !locator) {
+        // return `accum` to skip this link
+        return accum;
+      }
+
+      const fullId = [parentIds, id].filter(Boolean).join('.');
+
+      let calculatedHref = href; // allow consumer to pass href, but default to an href formed using the locator
+      if (locator) {
+        const storedLocator: Locator = locator; // never undefined
+        calculatedHref = locator.getRedirectUrl(locatorParams ?? {}); // NOTE: calculated href overwrites href from props
```

Comment:
Here is where the code should call `locator.getUrl` instead, which resolves with a user-friendly path string. This path string will be more useful to us in the code as we could check the `href` against URLs that fire from the history service, to check which nav item corresponds to the browser's current URL. 

The challenge currently is that `getUrl` is async, which imposes a lot of refactoring of this code and the component that uses this model.